README
Homework 1.03 - Path Finding
By: Edwin O. Martinez Velazquez

1.01:
Dungeon Generations randomly generates a dungeon in a 80*21 grid.
It will randomly generate between 5 to 7 rooms.
The rooms will be of a random size between 3x2 to 6x5. This is due to stability
It connects the rooms using Bresenham's Line Algorithm, making a line to the closest room, then looping the first room connected with the last one to force cycles.
Rooms, Rocks, and Corridors are their own structs.
The dungeon is a 80x21 matrix of a tagged union, containing the tag of type and an instance of room, rock, or corridor.

1.02:
Program now has --load & --save switch.
--load switch has optional parameter to specify which file to load from
--save switch has optional parameter to specify which file to save to
If no parameter for --load it will default to "dungeon". File must be a string, e.g., must contain at least 1 character other than a digit. 
If no parameter for --save it will default to "dungeon", unless --load switch has a specified file. In which case it will save to the file loaded from.
--load and --save can be used together, in which case it will load the file and then save it.
--load cannot be used with a seed. A digit only parameter after --load will be read as a seed and fail
--save can be used with a seed or random seed (no input argument)
File must be in the appropiate format, otherwise program file indicate it is not.
Saving file will indicate if there was any errors in the writing process.
Order of switches doesn't matter.

1.03:
Program now calculates distances from player.
A player is set randomly with the seed used for a random dungeon. For loaded dungeons, the player will be placed randomly.
Player is placed in any floor space (room or corridor or cell with a hardness of 0).
The program will generate the first dungeon as normal. Then it will graph the dungeon floors and perform Dijkstra's algorithm. This results in a cost map for non tunneler monsters.
The program will then generate a third map by graphing the entire dungeon and running Dijkstra algorithm on that. The resulting map is the cost for tunneler monsters.
Cost to each cell start from 0 (player's position) to 9, then a to z, and A to Z. If a cost is larger than Z (61) then the cell will be rendered normally.
The weight of each cell is determined by hardness. A hardness of 0 (floor) is 1. 
Cells with hardness of 1 to 84 also have a weight of 1, 85 to 170 have a weight of 2, and 171 to 254 have a weight of 3. Cells of 255 hardness have weights of 428400 (255 hardness x 1680 cells).
Tunnelling effort is not taken into account in this version as it is not required in the assignment specificaitons.
Borders are now draw with ~

1.04:
Game now has player and monsters.
Player is a drunken idiot. He moves randomly.
There are 16 types of monsters, combination of any 4 powers: intelligence, telepath, tunneler, and erratic. These powers correspond from least significant bit to most significant bit.
The monster is represented by the hexadecimal combinations of its powers, from 0 (no behaviour other than sight aggro) to f (all powers).
Intelligent monsters remember the last place they saw the player and move via shortest path to the known player position, straight line otherwise.
Telepaths always knows the player's location.
Tunnellers can go through rock, but have to chisel away 85 density points per turn
Erratic monsters have a 50% chance of just moving randomly instead.
Turns are determined via 100/speed with speed been a number between 5 and 20. This is queued through a priority queue.
New switch --nummon <int> is now available. This sets the number of monsters to spawn. Monsters are chosen at random and placed randomly, but following their rules. (non tunnelers will
only spawn inside the dungeon floor).

1.05:
Game now draws with curses in the scr window. It now also take key inputs to control the game:
Move up - w, k, 8, numpad 8
Move down - s, j, 2, numpad 2
Move left - a, h, 4, numpad 4
Move right - d, l, 6, numpad 6
Rest - Spacebar, center arrow, numpad 5
Go downstairs - >
Go upstairs - <
Display monster list - m
Scroll up through monster list - up arrow
Scroll down through monster list - down arrow
Exit monster list - esc
Save and exit game - S (upper case S, not to be confused with lower case s for move down)
The monster list is only scrollable if there are more monster than fits the screen.
A new dungeon is randomly generated every time you go up or down a stair. First dungeon is level 0 and has no upstairs. Every time you gown down a stair you reach another level.
To win you just clear all monsters in the current map. 
Player can no longer tunnel. 
Game is now colorful. Player is cyan, monsters are red, down stairs are black on green, up stairs are black on magenta, rooms are black on white, corridors are black on yellow.
Monster list is black on white, win screen is everything black on white, lose screen is everything black on red, save and exit is white on black.

1.06:
In this version monster and player have been converted from struct to c++ class. Character was also changed from struct to class and is a parent to both monster and player. 
CharacterController is a .cpp file. All references to monster, player, and character are wrapped in C functions for C++ interfacing in C. 
Fog of war has been implemented. The player can only see within a 3x3 grid. S/He will remember the dungeon as the player saw it and not as it changes.
The player can tell if a cell is within the visible 3x3 grid if the cell has colors, i.e., not white on black. The player can only see monsters within the grid as well.
NOTE: Without --nummon you automatically win the game. For some TAs the win/lose screen does not appear. Valgrind takes FOREVER and the invalid reads are index out of bounds error that the code handles safely.

1.07:
Added Monster parsing. Will now parse a file containing description of a monster.
Note: doesn't put them in the game yet. Creates a vector of monster class to store the data.
Need to use --parsemon flagg by itself

1.08:
Added object parsing. Will now parse files containing object. WARNING negative base numbers on dices are allowed only in objects, but not monsters. Only the base number can be negative.
With the --parsemon and --parseobj, it will parse and place the monsters and objects into the game respectively.
Bug fix from last iteration: It would skip a monster if it read a blank line. This was fixed in this version.
WARNING: Some monster are super fast now, good luck getting past level 0.

1.09:
Updated combat. Monsters use damage dice for attacking. No more automatic kills. Monsters can't kill other monsters. 
Updated UI to show monster's name and Health Points
Player can now pick up and use equipments. Currently only perks are damage dices per gear and speed bonus (or decrement if wearing a dress).
Updated UI for general gameplay. Health Points are displayed in the debug line on top. Player actions are the first line below the dungeon. Monster actions towards the player are in the line below that.
Updated UI to manage items and equipments. Now have an Inventory menu and an Equipment menu. Multiple commands to wear and take off equipment, as well as drop, remove, or destroy items.
When dropping items, they will fall in the floor. Stacked items are available and working. Everything resets when you go up or down a floor, so good luck making it far.
RECOMMENDATIONS: Use seed 1460517657. You will spawn next to the one ring and a stack of items. Lich Queen is in north east corner along with her prom dress and prada shoes.
Sorry, not sorry use flags: ./dungeonGenerator 1460517657 --nummon 10 --parsemon --parseobj  
NOTICE: I mapped some commands differently. As a PC gamer I reserve wasd for movements.
New Commands:
shift+w, (W) = Wear Item Dialog
t = take off equiped item
shift+d, (D) = drop item
x = expunge item
shift+i, (I) = Inventory List
shift+e, (E) = Equipment List
i = inspect item
All dialogs can be exit with Esc or Enter

1.10
Everything from 1.09 still applies.
In this version I made spells that can be fired as projectiles.
Spells are defined by making a SCROLL object in the object description file. 
Currently only offensive spells are implemented, thus the scroll object requires a damage dice to be parsed as a spell.
The attribute data in the scroll's object description defines mana cost.
It is also required that the name of the spell comes after "Scroll of", e.g., "Scroll of Ice".
The player can now learn up to 16 spells. Once learned they cannot be forgotten (unequipped) and the scroll is consumed.
To learn a spell you must have a spell scroll in your inventory, then press 'L' key and choose the scroll with keys '0' to '9'.
The spell will then be learned and appear in the Spell List, which can be viewed with 'M', not to be confused with monster list, which is 'm'.
To use a spell it must have been learned first. Then press 'A' to cast a spell by choosing it using keys 'a' to 'p'.
Once a valid spell has been chosen to cast, the user will be prompt with a direction. All valid movement keys can be used as direction and are displayed in this menu for you.
The spell will then launch away.
Fire and Earth spell (earth spell not included in my object description file) can break through rock and tunnel. All other spells cannot tunnel.
Spell will disappear once it hits a wall or monster. If it hits a monster it will receive damage (spell's damage calculated when fired).
All spell will cost mana, if mana hits 0 or less, then you can't cast anymore.
Only have Fire (tunneling) and Water (non tunneling) spells
WARNING: Known bug when casting a spall on a monster that is adjacent to the player, my spell vector gets a bit messed up, but it's handle by changing the supposed null spell color to default white on black.
RECOMMENDATIONS: Use seed 1460517657, e.g., ./dungeonGenerator 1460517657 --nummon 10 --parsemon --parseobj
New commands:
shift+m, (M) = Spell List
shift+l, (L) = Learn spell
shift+a, (A) = Cast spell

Usage for object parsing and item placing: ./dungeonGenerator --parseobj

Usage for monster parsing and placing: ./dungeonGenerator --parsemon

Usage using time as seed: ./dungeonGenerator

Usage using int as seed: ./dungeonGenerator <number>

Usage for load: ./dungeonGenerator --load <optional file name>

Usage for save: ./dungeonGenerator --save <optional file name> <optional seed>

Usage for load/save: ./dungeonGenerator --load <optional file name> --save <optional file name>

Usage for spawning monsters: ./dungeonGenerator --nummon <int>

CHANGELOG
1.01
1/21/16 Started Project
1/21/16 Made header with struct of rock, room, and corridor, along with an enumeration for each of those three
1/21/16		Also made a generic struct that can be one of the three
1/22/16 Made the union in the header
1/22/16 Added destroy
1/22/16 Made a union tag with a parent struct
1/22/16 Also better defined constructor and deconstructors
1/22/16 Fixed bugs in DungeonObjects.c
1/22/16 Fixed deconstructors bugs related to pointers
1/22/16 Change the structs inside union and structs to pointer
1/22/16 Removed pointers to structs
1/22/16 Added pass by reference in dungeonObjects.c
1/22/16 Trying pointer implementation again
1/22/16 Added pointer return to constructors
1/22/16 Removed pointers from header
1/22/16 Change from arrows to dots
1/22/16 Removed mallocs and deconstructors
1/22/16 Practicing rand() implementation and use from lecture
1/22/16 Practicing using modulos to get a specific range of random numbers
1/22/16 Testing my random implementation
1/22/16 Testing for loop with pointer instead of counter int
1/22/16 A lot of testing and bug fixing looping by pointer
1/22/16 After several hours finally figured out looping by pointers
1/23/16 Started Makefile
1/23/16 Beginning dungeonRoomGenerator to make array of random rooms of random sizes
1/23/16 Converting time_t to int
1/23/16 Fixed small pointer to int bug
1/23/16 Adding printf for debugging
1/23/16 Changed how I'm using random for room generator
1/23/16 Fixed parameters in header file
1/23/16 Changed loop termination for looping through pointer of rooms (array)
1/23/16 Experimenting loop termination with an invalid extra room at the end of the array to determine size
1/23/16 Fully fixed for loop termination
1/24/16 Testing tagged union
1/24/16 Bug fix in tagged union
1/24/16 Testing generating blank dungeonObjects
1/24/16 Fixed global double pointer to tagged union
1/24/16 Forgot stdio.h in dungeonLevelGenerator.can
1/24/16 Testing room placement
1/24/16 Fixed room generating in borders
1/24/16 Added more debugging printf
1/24/16 Fixed while loop condition for room placement
1/24/16 Changed drawing to switch case
1/24/16 Added corridor case for switch case
1/24/16 Fixed room drawing
1/24/16 Added condition to prevent room overlap
1/24/16 Fixed boolean condition
1/24/16 Added 1 spacing to room placement
1/24/16 Enhanced overlap check
1/24/16 Fixed boolean flags
1/24/16 Fixed 1 spacing
1/24/16 Started sorting rooms in rooms array by closest to first room, then closest to second, etc...
1/24/16 Bug fixing room sorting
1/24/16 Added debug printf to room sorting
1/24/16 Checking sorting results
1/24/16 Changed distance from center of room rather than corner
1/24/16 Implementing Bresenham's line algorithm
1/24/16 Bug fix to itoa()
1/24/16 itoa not supported, changing to snprintf
1/24/16 More use of debugging with printf
1/24/16 Experimenting with line algorithm
1/24/16 Experimenting enhancing the line algorithm to avoid diagonals
1/24/16 Succesfully implemented line algorithm without diagonals
1/24/16 Experimenting with more rooms and larger rooms to determine stability
1/24/16 Done experimenting. Most stable parameters are 5 to 7 rooms of up to size 6x5
1/24/16 Moved dungeon generation from main function to it's own function
1/24/16 Bug fixing final implementation
1/24/16 dungeon generator function returns a 2D array representation of dungeon and uses it to draw
1/26/16 Finished Makefile, README, and the code itself
1/26/16 Finished CHANGELOG. Ready to package and deliver

1.02
1/28/16 Started dungeonIO.h; Started header for saving and loading.
1/28/16 Implementing argument inputs
1/28/16 Added ctype.h and fix debug printf
1/28/16 looping from second argument
1/28/16 fixing isstring
1/29/16 Made dungeonIO.c and typdef Flags enumeration and implement contains flags
1/29/16 added dungeonIO to makefile and starting to test it in main
1/29/16 Fixed Makefile
1/29/16 Added debugging printf
1/29/16 changed line of debug printf
1/29/16 Testing flag array
1/29/16 bug fix
1/29/16 forgot semi colon
1/29/16 Should fix the flag array bug
1/29/16 Testing argument conditions
1/29/16 removed switch case
1/29/16 Checks if their is a seed and load flag
1/29/16 fixed checking for seed and load flag together
1/29/16 Finished all argument states
1/29/16 Forgot to return on invalid argument
1/29/16 Added density of 0 to room and corridor and change permanent rock from INT_MAX to 255
1/29/16 No permanent rock within borders
1/29/16 Testing getenv
1/29/16 testing slashes in string in getenv
1/29/16 testing getenv with strcat
1/29/16 testing checking for folder existing or not
1/29/16 Fixed typo in getenv
1/29/16 testing mkdir
1/29/16 forgot to include header for mkdir
1/29/16 testing null return from getenv
1/29/16 fixing segfault
1/29/16 More getenv testing
1/29/16 mkdir and getenv in one line
1/29/16 final testing, should work now
1/29/16 Cleaned up the debug code
1/29/16 removed if else for mkdir
1/29/16 Testing ssh
1/29/16 final testing of ssh key
1/29/16 removed bad use of getenv
1/29/16 Started load method
1/29/16 testing access
1/29/16 added errno.h
1/29/16 Some error fixing in access testing
1/29/16 Testing access with non existing file
1/29/16 added access() to load
1/29/16 testing strcat in --load
1/29/16 playing with string for --load
1/29/16 more playing with strings for --load
1/29/16 warning fix for playing with strings in --load
1/29/16 testing strdup in --load
1/29/16 was using strstr instead of strcat, bug fixed
1/29/16 --load will now check for file existance
1/29/16 testing --load param
1/29/16 testing input order
1/29/16 bug fix --load input
1/29/16 Fixing isstring()
1/29/16 more fixing isstring()
1/29/16 debugging isstring()
1/29/16 Fixed isstring()
1/29/16 --load parameters should be working
1/29/16 testing load()
1/29/16 debbuging warnings in load()
1/29/16 fixing warning in load()
1/29/16 trying freads
1/29/16 Trying closing file
1/29/16 debugging load()
1/29/16 looking for the error

Graph.cpp
#include "Graph.h"
#include <stdio.h>

graph_t internal_graph;
const vertex_t NULL_VERTEX = {.vertexData = {.space_type = ROCK, .space_union = {.rock = {.density = 255}}}, .weight =  428400, .x = -1, .y = -1};
const edge_t NULL_EDGE = {.source = {.vertexData = {.space_type = ROCK, .space_union = {.rock = {.density = 255}}}, .weight = 428400, .x = -1, .y = -1}, .target = {.vertexData = {.space_type = ROCK, .space_union = {.rock = {.density = 255}}}, .weight = 428400, .x = -1, .y = -1}, .weight = 428400};

graph_t Create_Graph()
{
	internal_graph.num_vertices = 0;
	internal_graph.num_edges = 0;
	internal_graph.vertices = (vertex_t *) malloc(sizeof(vertex_t));
	internal_graph.edges = (edge_t *) malloc(sizeof(edge_t));
	internal_graph.vertices[0] = NULL_VERTEX;
	internal_graph.edges[0] = NULL_EDGE;
	
	return internal_graph;
}

void Destroy_Graph(graph_t *destroy_graph)
{
	free(destroy_graph->vertices);
	free(destroy_graph->edges);
}

boolean Add_Vertex(Dungeon_Space_Struct cell, int x, int y)
{
	vertex_t new_vertex;
	new_vertex.vertexData = cell;
	switch(cell.space_type)
	{
		case ROCK:
			//new_vertex.weight = cell.space_union.rock.density;
			if(cell.space_union.rock.density <= 84)
			{
				new_vertex.weight = 1;
			}
			else if(cell.space_union.rock.density <= 170) 
			{
				new_vertex.weight = 2;
			}
			else if(cell.space_union.rock.density <= 254)
			{
				new_vertex.weight = 3;
			}
			else
			{
				new_vertex.weight = 428400;
			}
		break;
			
		
		case ROOM:
			new_vertex.weight = 1;
		break;
		
		case CORRIDOR:
			new_vertex.weight = 1;
		break;
	}
	new_vertex.x = x;
	new_vertex.y = y;
	
	internal_graph.num_vertices++;
	internal_graph.vertices = (vertex_t *) realloc(internal_graph.vertices, sizeof(vertex_t) + (sizeof(vertex_t) * internal_graph.num_vertices));
	internal_graph.vertices[internal_graph.num_vertices-1] = new_vertex;
	internal_graph.vertices[internal_graph.num_vertices] = NULL_VERTEX;
	
	return TRUE;
}

boolean Add_Edge(vertex_t source, vertex_t target)
{
	edge_t new_edge;
	new_edge.source = source;
	new_edge.target = target;
	new_edge.weight = (Compare_Vertices(target, NULL_VERTEX) == FALSE) ? target.weight : 428400;
	/*if(source.x == 0 && source.y == 0)
	{
		printf("Weight of edge at source [0][0] = %d\n", new_edge.weight);
	}*/
	internal_graph.num_edges++;
	internal_graph.edges = (edge_t *) realloc(internal_graph.edges, sizeof(edge_t) + (sizeof(edge_t) * internal_graph.num_edges));
	internal_graph.edges[internal_graph.num_edges-1] = new_edge;
	internal_graph.edges[internal_graph.num_edges] = NULL_EDGE;
	
	return TRUE;
}

vertex_t Get_Vertex(int x, int y)
{
	int v;
	for(v = 0; v < internal_graph.num_vertices; v++)
	{
		if((internal_graph.vertices[v].x == x) && (internal_graph.vertices[v].y == y))
		{
			return internal_graph.vertices[v];
		}
	}
	return NULL_VERTEX;
}

boolean Compare_Vertices(vertex_t key, vertex_t with)
{
	if((key.x == with.x) && (key.y == with.y))
	{
		return TRUE;
	}
	return FALSE;
}

boolean Contains_Vertex(vertex_t *vertex_collection, int num_elements, vertex_t find)
{
	int e;
	for(e = 0; e < num_elements; e++)
	{
		if(Compare_Vertices(vertex_collection[e], find) == TRUE)
		{
			return TRUE;
		}
	}
	return FALSE;
}

edge_t *Get_Edges_Of(vertex_t vertex)
{
	edge_t *adjacent_edges = (edge_t *) malloc(sizeof(edge_t) * 8);
	int e, edge_count = 0;
	for(e = 0; e < internal_graph.num_edges; e++)
	{
		edge_t current_edge = internal_graph.edges[e];
		if(Compare_Vertices(current_edge.source, vertex) == TRUE)
		{
			adjacent_edges[edge_count] = current_edge;
			edge_count++;
		}
	}
	
	return adjacent_edges;
}

graph_t GenerateGraph(Dungeon_Space_Struct **dungeon, boolean tunneler)
{
	Create_Graph();
	
	int y, x;
	for(y = 0; y < 21; y++)
	{
		for(x = 0; x < 80; x++)
		{
			Add_Vertex(dungeon[x][y], x, y);
		}
	}
	
	int v;
	for(v = 0; v < internal_graph.num_vertices; v++)
	{
		int e, a = -1, b = -1;
		for(e = 0; e < 8; e++)
		{
			if((internal_graph.vertices[v].x == 0) || (internal_graph.vertices[v].x == 79) || (internal_graph.vertices[v].y == 0) || (internal_graph.vertices[v].y == 20) || (tunneler == FALSE && internal_graph.vertices[v].vertexData.space_type == ROCK))
			{
				/*if(internal_graph.vertices[v].x == 0 && internal_graph.vertices[v].y == 0)
				{
					printf("Current is is [0][0] with a weight of %d\n", internal_graph.vertices[v].weight);
				}*/
				Add_Edge(internal_graph.vertices[v], NULL_VERTEX);
				//printf("[%d][%d] is a border\n", internal_graph.vertices[v].x, internal_graph.vertices[v].y);
			}
			else if(!(tunneler == FALSE && Get_Vertex(internal_graph.vertices[v].x + a, internal_graph.vertices[v].y + b).vertexData.space_type == ROCK))
			{
				/*if(Get_Vertex(internal_graph.vertices[v].x + a, internal_graph.vertices[v].y + b).x == 0 && Get_Vertex(internal_graph.vertices[v].x + a, internal_graph.vertices[v].y + b).y == 0)
				{
					printf("Neighboor is [0][0]\n");
				}*/
				Add_Edge(internal_graph.vertices[v], Get_Vertex(internal_graph.vertices[v].x + a, internal_graph.vertices[v].y + b));
				//printf("Vertex at [%d][%d] is of weight = %d\n", internal_graph.vertices[v].x+a, internal_graph.vertices[v].y+b, Get_Vertex(internal_graph.vertices[v].x + a, internal_graph.vertices[v].y + b).weight);
			}
			else
			{
				Add_Edge(internal_graph.vertices[v], NULL_VERTEX);
			}
			
			a++;
			if(a > 1)
			{
				a = -1;
				b++;
			}
			if(a == 0 && b == 0)
			{
				a = 1;
			}
		}
	}
	
	return internal_graph;
}